src/ndpp_ops.F90

```diff
@@ -0,0 +1,1083 @@
+
+  subroutine ndpp_tally_scatt_yn(this, i_nuclide, gin, score_index, &
+                                 filter_index, t_order, mult, is_analog, Ein, &
+                                 uvw, results, nuscatt)
+    type(Ndpp), intent(in) :: this      ! Ndpp object to act on
+    integer, intent(in) :: i_nuclide    ! index into nuclides array
+    integer, intent(in) :: gin          ! Incoming group index
+    integer, intent(in) :: score_index  ! dim = 1 starting index in results
+    integer, intent(in) :: filter_index ! dim = 2 starting index (incoming E filter)
+    integer, intent(in) :: t_order      ! # of scattering orders to tally
+    real(8), intent(in) :: mult         ! wgt or wgt * atom_density * flux
+    logical, intent(in) :: is_analog    ! Is this an analog or TL event?
+    real(8), intent(in) :: Ein          ! Incoming energy
+    real(8), intent(in) :: uvw(3)       ! direction coordinates
+    type(TallyResult), intent(inout) :: results(:,:) ! Tally results storage
+    logical, optional, intent(in) :: nuscatt         ! Is this for nuscatter?
+
+    integer :: g         ! outgoing energy group index
+    integer :: g_filter  ! outgoing energy group index
+    integer :: i_score   ! index of score dimension of results
+    integer :: l         ! legendre moment index
+    real(8) :: norm      ! Interpolation constant, multiplied by sigS (if in TL)
+    integer :: gmin      ! Minimum group transfer in ndpp_outgoing
+    integer :: gmax      ! Maximum group transfer in ndpp_outgoing
+    integer :: num_lm    ! Number of m for this l
+    real(8) :: sigs_el   ! Elastic x/s
+    real(8) :: sigs_inel ! Inelastic x/s
+    logical :: nuscatt_flag
+
+    if (present(nuscatt)) then
+      nuscatt_flag = nuscatt
+    else
+      nuscatt_flag = .false.
+    end if
+
+    ! First we need to calculate the x/s to use when creating our distribution.
+    call calc_scatter_xs(i_nuclide, sigs_el, sigs_inel)
+
+    ! Now combine the elastic and inelastic distributions to one set,
+    ! which will be in ndpp_outgoing(thread_id,:,:)
+    call generate_ndpp_distrib_pn(this, nuscatt_flag, gin, t_order, Ein, sigs_el, sigs_inel, &
+                                  norm, gmin, gmax)
+
+    ! Apply mult to the n ormalization constant, norm
+    norm = norm * mult
+    ! Now apply sigS if we are in tracklength mode
+    if (.not. is_analog) then
+      norm = norm * (micro_xs(i_nuclide) % total - &
+        micro_xs(i_nuclide) % absorption)
+    end if
+
+    ! Add the combined distribution to our tally for each order
+    do g = gmin, gmax
+      g_filter = filter_index + g - 1
+      i_score = score_index
+      do l = 0, t_order
+        num_lm = 2 * l + 1
+!$omp critical
+        results(i_score: i_score + num_lm - 1, g_filter) % value = &
+          results(i_score: i_score + num_lm - 1, g_filter) % value + &
+          ndpp_outgoing(thread_id, l + 1, g) * norm * calc_rn(l, uvw)
+!$omp end critical
```

Comment:
Almost, `atomic` is reserved for instructions which processors have specific instructions for and thus you usually only see them for very simple changes like `i = i + 1`.  Critical is for everything else.  Since this block has array slicing, multiple multiplications and additions, I kept it critical.

However.... in making sure I was remembering correctly, I did a google search on your question.  Well, I found out that `omp critical` sections should be named, otherwise they are all treated as the same critical question.  I did not know that.  So, now all of these critical sections are named per my last commit which I will make once finished addressing your comments. I will also create a PR to do the same for the rest of OpenMC.